src/mapping/ginkgo_kernels/GinkgoRBFKernels.cpp

```diff
@@ -0,0 +1,188 @@
+          mtx[i * rowLength + j] = 0;
+        }
+
+#ifdef __NVCC__
+
+        // Use this as readonly shared buffer
+        __shared__ double prefetchedEvalPoint[3];
+        double            y;
+
+        // Check if current block is at the end of a matrix row and induces a line break
+        // However, if a block is larger than an entire row, we need to disable it since we now can't make sure
+        // it does not still span across two lines.
+        // We have to use uint64_t because these matrices become so large that using this
+        // if-condition overflows with int32
+        uint64_t blockID        = blockIdx.x;
+        uint64_t leftThreadNum  = blockID * blockDim.x;
+        uint64_t rightThreadNum = (blockID + 1) * blockDim.x - 1;
+        if (blockDim.x >= rowLength || (blockDim.x < rowLength && leftThreadNum % rowLength > rightThreadNum % rowLength)) {
+
+          // Since this block spans across two lines, we have to use global memory and cannot use prefetched memory
+          for (size_t k = 0; k < dataDimensionality; ++k) {
+            y    = supportPoints[k * inputRowLength + j] - targetPoints[k * outputRowLength + i];
+            dist = fma(y, y, dist);
+          }
+        } else {
+
+          // If this block is indeed only in one row, we can make thread 0 in each block responsible for prefetching values into shared memory
+          if (0 == threadIdx.x) {
+            prefetchedEvalPoint[0] = targetPoints[i];
+            prefetchedEvalPoint[1] = targetPoints[outputRowLength + i];
+            prefetchedEvalPoint[2] = targetPoints[2 * outputRowLength + i];
+          }
+          // Let all threads in a block wait until memory is prefetched
+          __syncthreads();
+
+          for (size_t k = 0; k < dataDimensionality; ++k) {
+            y    = supportPoints[k * inputRowLength + j] - prefetchedEvalPoint[k]; // <- This should not impose a bank conflict since CUDA offers broadcasting if a warp accesses the same shared memory adress
+            dist = fma(y, y, dist);
+          }
+        }
+
+        dist = sqrt(dist);
+
+#else
+        const unsigned int supportPointOffset = dataDimensionality * j; // Point of current column
+        const unsigned int targetPointOffset  = dataDimensionality * i; // Point of current row
+        // Loop over each dimension and calculate euclidian distance
+        for (size_t k = 0; k < dataDimensionality; ++k) {
+          dist += std::pow(supportPoints[supportPointOffset + k] - targetPoints[targetPointOffset + k], 2) * static_cast<int>(activeAxis.at(k));
+        }
+
+        dist                                  = std::sqrt(dist);
+#endif
+
+        mtx[i * rowLength + j] = f(dist, rbf_params);
+      },
+      gko::dim<2>{n1, n2}, n2, dataDimensionality, activeAxis, mtx, supportPoints, targetPoints, f, rbf_params, inputRowLength, outputRowLength, addPolynomial, extraDims);
+}
+
+// Here, we need to instantiate all possible variants for each basis function
```

Comment:
I understand that this is necessary by the way Ginkgo works. Internally, they often use macro expansions for such instantiations. I would also consider doing this here as editing the kernel function signature will become very tedious. However, feel free to decide this on your own, just a comment.

Reply:
What about having this function signature as `#define` macro which takes one input which is the name of the RBF class?